Lib/test/test_ctypes/test_c_simple_type_meta.py

```diff
@@ -85,3 +85,59 @@ class Sub(CtBase):
 
         self.assertIsInstance(POINTER(Sub), p_meta)
         self.assertTrue(issubclass(POINTER(Sub), Sub))
+
+    def test_creating_pointer_in_dunder_init_1(self):
+        # Test for modern metaclass initialization that does not
+        # require recursion avoidance.
+
+        class ct_meta(type):
+            def __init__(self, name, bases, namespace):
```

Comment:
- Please don't leave out the `super().__init__` call when subclassing.
- Please do limit recursion, as with `__new__`.
```suggestion
        class ct_meta(type):
            def __init__(self, name, bases, namespace):
                super().__init__(name, bases, namespace)

                # Avoid recursion.
                # (See test_creating_pointer_in_dunder_new_1)
                if bases == (c_void_p,):
                    return
                if issubclass(self, PtrBase):
                    return
```

(And similar in `test_creating_pointer_in_dunder_init_2`.)

The reason you don't need recursion avoidance here is that the internal `PyCSimpleType` is not set up for diamond inheritance patterns, and doesn't call `super().__init__` (that is, `ct_meta.__init__` here). That's arguably a bug. (It definitely would be one if we wanted to support such subclassing.)
If you switch the metaclass bases, writing `class p_meta(ct_meta, PyCSimpleType)`, recursion avoidance becomes necessary again.

Reply:
Good catch! thank you.